src/proxy/http/profiles.rs

```diff
@@ -0,0 +1,389 @@
+    labels: Arc<IndexMap<String, String>>,
+    response_classes: Arc<Vec<ResponseClass>>,
+}
+
+#[derive(Clone, Debug)]
+pub enum RequestMatch {
+    All(Vec<RequestMatch>),
+    Any(Vec<RequestMatch>),
+    Not(Box<RequestMatch>),
+    Path(Regex),
+    Method(http::Method),
+}
+
+#[derive(Clone, Debug)]
+pub struct ResponseClass {
+    is_failure: bool,
+    match_: ResponseMatch,
+}
+
+#[derive(Clone, Debug)]
+pub enum ResponseMatch {
+    All(Vec<ResponseMatch>),
+    Any(Vec<ResponseMatch>),
+    Not(Box<ResponseMatch>),
+    Status {
+        min: http::StatusCode,
+        max: http::StatusCode,
+    },
+}
+
+// === impl Route ===
+
+impl Route {
+    pub fn new<I>(label_iter: I, response_classes: Vec<ResponseClass>) -> Self
+    where
+        I: Iterator<Item = (String, String)>,
+    {
+        let labels = {
+            let mut pairs = label_iter.collect::<Vec<(String, String)>>();
+            pairs.sort_by(|(k0, _), (k1, _)| k0.cmp(k1));
+            Arc::new(IndexMap::from_iter(pairs))
+        };
+
+        Self {
+            labels,
+            response_classes: response_classes.into(),
+        }
+    }
+
+    pub fn labels(&self) -> &Arc<IndexMap<String, String>> {
+        &self.labels
+    }
+
+    pub fn response_classes(&self) -> &Arc<Vec<ResponseClass>> {
+        &self.response_classes
+    }
+}
+
+// === impl RequestMatch ===
+
+impl RequestMatch {
+    fn is_match<B>(&self, req: &http::Request<B>) -> bool {
+        match self {
+            RequestMatch::Method(ref method) => req.method() == *method,
+            RequestMatch::Path(ref re) => re.is_match(req.uri().path()),
+            RequestMatch::Not(ref m) => !m.is_match(req),
+            RequestMatch::All(ref matches) => {
+                for ref m in matches {
+                    if !m.is_match(req) {
+                        return false;
+                    }
+                }
+                true
+            }
+            RequestMatch::Any(ref matches) => {
+                for ref m in matches {
+                    if m.is_match(req) {
+                        return true;
+                    }
+                }
+                false
+            }
+        }
+    }
+}
+
+// === impl ResponseClass ===
+
+impl ResponseClass {
+    pub fn new(is_failure: bool, match_: ResponseMatch) -> Self {
+        Self { is_failure, match_ }
+    }
+
+    pub fn is_failure(&self) -> bool {
+        self.is_failure
+    }
+
+    pub fn is_match<B>(&self, req: &http::Response<B>) -> bool {
+        self.match_.is_match(req)
+    }
+}
+
+// === impl ResponseMatch ===
+
+impl ResponseMatch {
+    fn is_match<B>(&self, req: &http::Response<B>) -> bool {
+        match self {
+            ResponseMatch::Status { ref min, ref max } => {
+                *min <= req.status() && req.status() <= *max
+            }
+            ResponseMatch::Not(ref m) => !m.is_match(req),
+            ResponseMatch::All(ref matches) => {
+                for ref m in matches {
```

Comment:
Again, TIOLI, but these for-loops would be terser as `Iterator::all`/`Iterator::any`...